dragan2234/worldcoin-scroll-bridge
# create a relayer service that periodically calls propagateRoot()

Resolves #9

## Description
This PR implements a Task Manager service to sync the World ID Merkle tree between the Scroll World ID contract and the Mainnet World ID contract. The service ensures consistency between these two blockchain environments by comparing the latest Merkle tree roots and propagating updates when necessary.

## Changes Made / Features Added
1. Implemented a Task Manager service that:
   - Continuously monitors both Scroll and Mainnet World ID contracts
   - Compares the latest Merkle tree roots from both contracts
   - Triggers a Propagate Root task when discrepancies are detected

2. Added a Propagate Root task that:
   - Generates a transaction to update the Scroll World ID root
   - Utilizes a relayer service to ensure the transaction is mined

3. Created a sync state endpoint to provide real-time synchronization status

4. Implemented a `/serviceStatus` endpoint to return the server status

5. Added configuration options for:
   - Network providers (L1 and L2)
   - Relayer service (tx-sitter)
   - Database connection
   - Server address and port

6. Included detailed setup instructions for local development, including:
   - Prerequisites installation
   - Docker setup for the database
   - Local node setup
   - Contract deployment
   - Relayer service setup (tx-sitter and OZ Relayer)

7. Provided a sample `config.toml` file for easy service configuration

## How to Test
1. Follow the "GETTING STARTED" section in the readme to set up the local development environment
2. Configure the service using the provided `config.toml` template
3. Run the service using `RUST_LOG=info cargo run config.toml`
4. Test the sync state and service status endpoints
5. Verify that the service correctly detects and propagates root updates between contracts

## Additional Notes
- The readme file has been updated to reflect the new service implementation and setup process.
- A Comprehensive video explaining how the service works and how to set it up with the link below.

https://drive.google.com/file/d/1SxaeIsdaFyXHy474axXdNad4rRzB3fYW/view?usp=sharing


## Checklist
- [X] Updated documentation (readme)
- [x] Ensured CI test suite passes (including lints)
- [x] Reviewed and added any necessary exemptions for cargo vet
